Docs: Device Trust role and locking support (#21915)
gravitational/teleport: docs/pages/access-controls/guides/device-trust.mdx

user/device combination.

<Admonition type="tip" title="Enrollment">
  Device enrollment is key in correctly identifying a device throughout its
  lifetime. The more trusted the enrollment operator, the better the ongoing

>   guarantees that the device itself is trustworthy.

Reviewer: I see someone else beat me to it. Still adding this comment so you can just commit the suggestion. Same applies to comments below.

spec:
  allow:
    rules:

>       - resources: ['device']

Reviewer 2: all of these are auto applied by my vscode which picks up from local prettierrc, eslint, editorconfig etc.  Let me check this again.

>       - resources: ['device']

Reviewer 3: Please do, this adds noise to both the PR and file history without a real benefit behind it.

>       - resources: ['device']

Reviewer: In this case, I'd advocate for keeping the change. I understand that you don't _need_ an indent for lists under keys in yaml, but IMO it makes it a lot easier to look at the document and parse the parent/children.

>       - resources: ['device']

Reviewer: In this case, I'd advocate for keeping the change. I understand that you don't need an indent for lists under keys in yaml, but IMO it makes it a lot easier to look at the document and parse the parent/children.

The device above is now a trusted device.

## Optional: Cluster-wide device enforcement

> Cluster-wide configuration enforces trusted device access at cluster level. It can be configured with the following three modes of operation, represented by the
> `device_trust.mode` authentication setting:

Reviewer: Hard break around 80 chars, and minor grammar edits:

```suggestion
Cluster-wide configuration enforces trusted device access at the cluster level.
It can be configured with the following three modes of operation, represented
by the `device_trust.mode` authentication setting:
```

The device above is now a trusted device.

> ## Trusted device access

Reviewer 2: I can update the heading to "Enforce Trusted Access" but I don't think we should write "Optional". What's the point of enrolling a device if not for having authenticated device access? 

"Optional" should go for steps that adds extra features without which, the core feature would still work fine. Here Enrolling Device are setup steps and Authenticated device is the core feature.

> ## Trusted device access

Reviewer 2: For now, I've rolled back headings to align with current preference. We can always argue about styles later :)

> ## Trusted device access

Reviewer 3: The feature is not useless without device enforcement, but you do have a point. Making role-based enforcement "Step 3" seems like a good idea.

  Additionally, it requires a trusted device for all SSH, Database and
  Kubernetes connections.

<Admonition type="warning" title="Web UI">
  The Web UI is not capable of trusted device access during the device trust
  preview. Only `tsh` and Teleport Connect are able to fulfill device mode
  `required`.
</Admonition>

### Per-Role

> To enforce authenticated device checks for a specific role, update the role as the following:

Reviewer 2: @codingllama, I've updated reference to `require_session_mfa`. However I could not find any reference to `require_mfa_type` in our docs?

> To enforce authenticated device checks for a specific role, update the role as the following:

Reviewer 3: Nevermind about RequireMFAType - it looks like it's a field in our protos, which is where I got it from, but not in the user-visible configs.

```diff
kind: role
version: v6
metadata:
  name: example-role
spec:
  options:
    # require authenticated device check for this role

> +   device_trust_mode: "required" # add this line

Reviewer 3: Do we have a reference page for role options? Let's update it if we do. (Maybe @alexfornuto knows?)

Reviewer 4: Yes, we do. Please update `role-spec.mdx` and `access-controls/reference.mdx`.

Reviewer 2: Resolved in the following commit.

  allow:
    ...
  deny:
    ...

```

<Admonition type="tip" title="Per-role vs. Cluster-wide">
Per-role configuration let's progressively block unauthorized device access and is useful for getting started with Teleport Device Trust.

Once all the devices used by engineers are enrolled to Teleport, cluster-wide configuration will ensure every resources can only ever be accessed with a trusted device.

> 

Reviewer 2: It's valid but not recommended security wise. Cluster-wide configuration is a failsafe mode, even if there's role misconfiguration, one can be sure that device authentication would be applied.

It's similar case if you compare with 2FA. Having 2FA enabled by default for all is better than conditional 2FA. Of course, it may not be feasible for every use cases and this is where granular policies step in to create exceptional cases. Otherwise, we should advocate for globally applied rules.

accesses.

<Admonition type="warning" title="Web UI">

>   The Web UI is not capable of trusted device access during the device trust
>   preview. Only `tsh` and Teleport Connect are able to fulfill device mode
>   `required`.

Reviewer: ```suggestion
The Web UI is not capable of trusted device access during the device trust
preview. Only `tsh` and Teleport Connect are able to fulfill device mode
`required`.
```

$ tctl create -f cap.yaml
cluster auth preference has been updated
```

> 

Reviewer: Good opportunity to use the new `tctl edit` feature:

```suggestion

You can also edit this configuration directly:

```code
$ tctl edit cluster_auth_preference
```

```

Reviewer 2: Thanks for this, I didn't knew we have `tctl edit` command.

    device_trust:
+     mode: "required" # add this line
```

> 

Reviewer: Disagree. Having a newlines surrounding code blocks is encouraged.

```

<Admonition type="tip" title="Trusted Clusters">

>   It is possible to use [trusted
>   clusters](../../management/admin/trustedclusters.mdx) to limit the impact of
>   device mode `required`. A leaf cluster in mode `required` will enforce access
>   to all of its resources, without imposing the same restrictions to the root
>   cluster. Likewise, a root cluster will not enforce device trust on resources
>   in leaf clusters.

Reviewer: ```suggestion
It is possible to use [trusted
clusters](../../management/admin/trustedclusters.mdx) to limit the impact of
device mode `required`. A leaf cluster in mode `required` will enforce access
to all of its resources, without imposing the same restrictions to the root
cluster. Likewise, a root cluster will not enforce device trust on resources
in leaf clusters.
```

leaf clusters.
</Admonition>

## Optional: Role-based device enforcement

> Role-based configuration enforces trusted device access at role level. It can be configured with `device_trust_mode` (`required`, `optional`, `off`) option and applies to the resources in its `allow` rules. It works similar to [`require_session_mfa`](per-session-mfa.mdx).

Reviewer: ```suggestion
Role-based configuration enforces trusted device access at the role level. It
can be configured with the `spec.options.device_trust_mode` (`required`,
`optional`, `off`) option and applies to the resources in its `allow` rules. It
works similar to [`require_session_mfa`](./per-session-mfa.mdx).
```

> To enforce authenticated device checks for a specific role, update the role as the following:

Reviewer: ```suggestion
To enforce authenticated device checks for a specific role, update the role with the following:
```

```diff
kind: role
version: v6
metadata:
  name: example-role
spec:
  options:
    # require authenticated device check for this role
+   device_trust_mode: "required" # add this line
  allow:
    logins: ['admin']
    kubernetes_groups: ['edit']
    node_labels:
      '*': '*'
    ...

```

<Admonition type="tip" title="Per-role vs. Cluster-wide">

> Per-role configuration let's progressively block unauthorized device access and is useful for getting started with Teleport Device Trust.

Reviewer: ```suggestion
Per-role configuration let's you progressively block unauthorized device access,
and is useful for getting started with Teleport Device Trust.
```

> Once all the devices used by engineers are enrolled to Teleport, cluster-wide configuration will ensure every resources can only ever be accessed with a trusted device.

Reviewer: ```suggestion
Once all the devices used by engineers are enrolled in Teleport, cluster-wide
configuration will ensure every resources can only ever be accessed with a
trusted device.
```

</Admonition>

## Optional: Locking a trusted device

> Similar to [session and identity locking](./locking.mdx), a device can be locked using `tctl lock`.
> 
> Locking blocks certificate issuance and ongoing or future accesses originating from a locked device.
> Locking device only works if device trust is enabled and if the device is enrolled to Teleport.

Reviewer: ```suggestion
Similar to [session and identity locking](./locking.mdx), a device can be locked
using `tctl lock`.

Locking blocks certificate issuance and ongoing or future accesses originating
from a locked device. Locking a device only works if device trust is enabled and
if the device is enrolled to Teleport.
```

> Find device ID to lock:

Reviewer: ```suggestion
Find a device ID to lock:
```

```code
$ tctl devices ls
Asset Tag     OS    Enroll Status   Device ID
------------ -----  ------------- ------------------------------------
(=devicetrust.asset_tag=)  macOS  enrolled      9cdfc0ad-64b7-4d9c-9342-50e97f418ba0
```

> Lock device:

Reviewer: ```suggestion
Lock a device:
```

```code
$ tctl lock --device 9cdfc0ad-64b7-4d9c-9342-50e97f418ba0 --ttl=12h
Created a lock with name "5444970a-39a0-4814-968d-e58b4a8fa686".
```

> Now if when user tries to access, say an SSH server, Teleport will deny access.

Reviewer: ```suggestion
Now, if a user on that device tries to access an SSH server for example, Teleport will deny access:
```

leaf clusters.
</Admonition>

## Optional: Role-based trusted device enforcement

Role-based configuration enforces trusted device access at the role level. It
can be configured with the `spec.options.device_trust_mode` (`required`,
`optional`, `off`) option and applies to the resources in its `allow` rules. It
works similar to [`require_session_mfa`](./per-session-mfa.mdx).

To enforce authenticated device checks for a specific role, update the role with the following:

```diff
kind: role
version: v6
metadata:
  name: example-role
spec:
  options:
    # require authenticated device check for this role
+   device_trust_mode: "required" # add this line
  allow:
    logins: ['admin']
    kubernetes_groups: ['edit']
    node_labels:
      '*': '*'
    ...

```

<Admonition type="tip" title="Per-role vs. Cluster-wide">
Per-role configuration let's you progressively block unauthorized device access,
and is useful for getting started with Teleport Device Trust.

> Once all the devices used by engineers are enrolled in Teleport, cluster-wide

Reviewer 3: Suggestion: drop the assumption about "engineers", also makes the sentence shorter.

```suggestion
Once all the devices used are enrolled in Teleport, cluster-wide
```

configuration will ensure every resources can only ever be accessed with a
trusted device.
</Admonition>

> ## Optional: Locking a trusted device

Reviewer 4: I disagree with saying "trusted device" everywhere. _Locking a device_ is just as clear and is 25% less words. Teleport doesn't manage "untrusted devices" so I don't think this qualifier is necessary.

Reviewer 2: Agree with that. Will remove "trusted" here.

Similar to [session and identity locking](./locking.mdx), a device can be locked
using `tctl lock`.

Locking blocks certificate issuance and ongoing or future accesses originating
from a locked device. Locking a device only works if device trust is enabled and
if the device is enrolled to Teleport.

Find a device ID to lock:

```code
$ tctl devices ls
Asset Tag     OS    Enroll Status   Device ID
------------ -----  ------------- ------------------------------------

> (=devicetrust.asset_tag=)  macOS  enrolled      9cdfc0ad-64b7-4d9c-9342-50e97f418ba0

Reviewer 3: I would if we should create a variable for the device ID, for consistency sake? WDYT?

Reviewer 2: I think thats a good idea. Updating this one with variable.

$ tctl create -f cap.yaml
cluster auth preference has been updated
```

> 

Reviewer: Disagree. Having a newlines surrounding code blocks is encouraged.

```

<Admonition type="tip" title="Trusted Clusters">
  It is possible to use [trusted
  clusters](../../management/admin/trustedclusters.mdx) to limit the impact of
  device mode `required`. A leaf cluster in mode `required` will enforce access
  to all of its resources, without imposing the same restrictions to the root
  cluster. Likewise, a root cluster will not enforce device trust on resources
  in leaf clusters.
</Admonition>

## Locking device

> Similar to [session and identity locking](./locking.mdx), a device can be locked using `tctl lock`.

Reviewer 3: We should explain here what locking a device means and how it takes effect. Teleport doesn't identify trusted devices, so if device trust is disabled or a device is unenrolled, locking it will have no effect. This can be confusing for users, so let's spell  it out.

There are some details on https://github.com/gravitational/teleport/pull/21667 too.

Reviewer 4: Please update `locking.mdx` too  (list a device as a valid lock target)

Reviewer 2: Resolved in following commit. Added device lock target in `locking.mdx` as well.

```code
# Find device id to lock
$ tctl devices ls
Asset Tag     OS    Enroll Status   Device ID
------------ -----  ------------- ------------------------------------
C0XXXXXXXXXX  macOS  enrolled      9cdfc0ad-64b7-4d9c-9342-50e97f418ba0

# Lock device

> $ tctl lock --device 9cdfc0ad-64b7-4d9c-9342-50e97f418ba0 --ttl=12h

Reviewer 3: Please update the tctl reference by adding the `--device` flag.

Reviewer 2: Can you please point the line number again? Above command reference already uses `--device` flag?

Reviewer 3: I was thinking of https://goteleport.com/docs/reference/cli/#tctl, but it looks like `tctl lock` is not there. I think you got it on locking.mdx, though.

Reviewer 4: Let's add `tctl lock` to that page. Doesn't have to be in this PR, but we can at least create a new issue to track that.
